<a class="anchor" id="0"></a>
# **Home Work 1: kNN Classifier**


### 제출마감: 4월 22일
### 배점: 5점

### 10문제(코드 1 - 코드 10)를 완성하세요.

### **10문제중 5문제 이상 정답이면, 5점 만점 입니다.**


In [1]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns 
%matplotlib inline


In [2]:
import warnings

warnings.filterwarnings('ignore')

# **1. Import dataset** 

In [3]:
# data.csv를 읽어들이세요. header=None 옵션을 사용하세요.

# 코드 1

,0,1,2,3,4,5,6,7,8,9,10
0,1000025,5,1,1,1,2,1,3,1,1,2
1,1002945,5,4,4,5,7,10,3,2,1,2
2,1015425,3,1,1,1,2,2,3,1,1,2
3,1016277,6,8,8,1,3,4,3,7,1,2
4,1017023,4,1,1,3,2,1,3,1,1,2
...,...,...,...,...,...,...,...,...,...,...,...
685,534555,1,1,1,1,2,1,1,1,1,2
686,536708,1,1,1,1,2,1,1,1,1,2
687,566346,3,1,1,1,2,1,2,3,1,2
688,603148,4,1,1,1,2,1,1,1,1,2


# **2. Exploratory data analysis** 

In [53]:
# view dimensions of dataset

# 코드 2

We can see that there are 690 instances and 11 attributes in the data set. 


In the dataset description, it is given that there are 10 attributes and 1 `Class` which is the target variable. So, we have 10 attributes and 1 target variable.

### View top 5 rows of dataset

In [5]:
# preview the dataset

# 코드 3

,0,1,2,3,4,5,6,7,8,9,10
0,1000025,5,1,1,1,2,1,3,1,1,2
1,1002945,5,4,4,5,7,10,3,2,1,2
2,1015425,3,1,1,1,2,2,3,1,1,2
3,1016277,6,8,8,1,3,4,3,7,1,2
4,1017023,4,1,1,3,2,1,3,1,1,2


### Rename column names

We can see that the dataset does not have proper column names. We should give proper names to the columns. I will do it as follows:-

In [6]:
col_names = ['Id', 'Clump_thickness', 'Uniform_Cell_Size', 'Uniform_Cell_Shape', 'Marginal_Adhesion', 
             'Single_Epithelial_Cell_Size', 'Bare_Nuclei', 'Bland_Chromatin', 'Normal_Nucleo', 'Mitoses', 'Class']

# 코드 4

df.columns

Index(['Id', 'Clump_thickness', 'Uniform_Cell_Size', 'Uniform_Cell_Shape',
       'Marginal_Adhesion', 'Single_Epithelial_Cell_Size', 'Bare_Nuclei',
       'Bland_Chromatin', 'Normal_Nucleo', 'Mitoses', 'Class'],
      dtype='object')

We can see that the column names are renamed. Now, the columns have meaningful names.

In [7]:
# let's agian preview the dataset

df.head()

,Id,Clump_thickness,Uniform_Cell_Size,Uniform_Cell_Shape,Marginal_Adhesion,Single_Epithelial_Cell_Size,Bare_Nuclei,Bland_Chromatin,Normal_Nucleo,Mitoses,Class
0,1000025,5,1,1,1,2,1,3,1,1,2
1,1002945,5,4,4,5,7,10,3,2,1,2
2,1015425,3,1,1,1,2,2,3,1,1,2
3,1016277,6,8,8,1,3,4,3,7,1,2
4,1017023,4,1,1,3,2,1,3,1,1,2


### Drop unnecessary columns


Here, `Id` is the unnecessary column. So, I will drop it first.

In [8]:
# drop Id column from dataset

# 코드 5

### View summary of dataset


In [9]:
# view summary of dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 10 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   Clump_thickness              690 non-null    int64 
 1   Uniform_Cell_Size            690 non-null    int64 
 2   Uniform_Cell_Shape           690 non-null    int64 
 3   Marginal_Adhesion            690 non-null    int64 
 4   Single_Epithelial_Cell_Size  690 non-null    int64 
 5   Bare_Nuclei                  690 non-null    object
 6   Bland_Chromatin              690 non-null    int64 
 7   Normal_Nucleo                690 non-null    int64 
 8   Mitoses                      690 non-null    int64 
 9   Class                        690 non-null    int64 
dtypes: int64(9), object(1)
memory usage: 54.0+ KB


We can see that the `Id` column has been removed from the dataset. 

We can see that there are 9 numerical variables and 1 categorical variable in the dataset. 

### Frequency distribution of values in variables

In [10]:
for var in df.columns:
    
    print(df[var].value_counts())


1     144
5     128
3     105
4      78
10     69
2      49
8      46
6      34
7      23
9      14
Name: Clump_thickness, dtype: int64
1     379
10     65
3      52
2      45
4      40
5      30
8      27
6      27
7      19
9       6
Name: Uniform_Cell_Size, dtype: int64
1     348
2      59
10     56
3      56
4      44
5      34
7      30
6      29
8      27
9       7
Name: Uniform_Cell_Shape, dtype: int64
1     403
2      58
3      56
10     55
4      32
8      25
6      22
5      21
7      13
9       5
Name: Marginal_Adhesion, dtype: int64
2     382
3      70
1      47
4      46
6      41
5      39
10     31
8      21
7      11
9       2
Name: Single_Epithelial_Cell_Size, dtype: int64
1     398
10    132
2      29
5      28
3      27
8      21
4      18
?      16
9       9
7       8
6       4
Name: Bare_Nuclei, dtype: int64
3     165
2     165
1     148
7      73
4      39
5      34
8      27
10     18
9      11
6      10
Name: Bland_Chromatin, dtype: int64
1     438
10     60
3  

The result shows that data type of `Bare_Nuclei` is of type integer. But the summary of the dataframe shows that it is type object. So, I will explicitly convert its data type to integer.

### Convert data type of Bare_Nuclei to integer

In [11]:
df['Bare_Nuclei'] = pd.to_numeric(df['Bare_Nuclei'], errors='coerce')

### Check data types of columns of dataframe

In [12]:
df.dtypes

Clump_thickness                  int64
Uniform_Cell_Size                int64
Uniform_Cell_Shape               int64
Marginal_Adhesion                int64
Single_Epithelial_Cell_Size      int64
Bare_Nuclei                    float64
Bland_Chromatin                  int64
Normal_Nucleo                    int64
Mitoses                          int64
Class                            int64
dtype: object

Now, we can see that all the columns of the dataframe are of type numeric.

### Summary of variables


- There are 10 numerical variables in the dataset.



- The first 9 variables are feature variables and last variable `Class` is the target variable.




### Explore problems within variables


Now, I will explore problems within variables.


### Missing values in variables

In [13]:
# check missing values in variables

df.isnull().sum()

Clump_thickness                 0
Uniform_Cell_Size               0
Uniform_Cell_Shape              0
Marginal_Adhesion               0
Single_Epithelial_Cell_Size     0
Bare_Nuclei                    16
Bland_Chromatin                 0
Normal_Nucleo                   0
Mitoses                         0
Class                           0
dtype: int64

We can see that the `Bare_Nuclei` column contains missing values. We need to dig deeper to find the frequency distribution of 
values of `Bare_Nuclei`.

In [14]:
# check `na` values in the dataframe

df.isna().sum()

Clump_thickness                 0
Uniform_Cell_Size               0
Uniform_Cell_Shape              0
Marginal_Adhesion               0
Single_Epithelial_Cell_Size     0
Bare_Nuclei                    16
Bland_Chromatin                 0
Normal_Nucleo                   0
Mitoses                         0
Class                           0
dtype: int64

We can see that the `Bare_Nuclei` column contains 16 `nan` values.

In [15]:
# check frequency distribution of `Bare_Nuclei` column

# 코드 6

1.0     398
10.0    132
2.0      29
5.0      28
3.0      27
8.0      21
4.0      18
9.0       9
7.0       8
6.0       4
Name: Bare_Nuclei, dtype: int64

In [16]:
# check unique values in `Bare_Nuclei` column

# 코드 7

array([ 1., 10.,  2.,  4.,  3.,  9.,  7., nan,  5.,  8.,  6.])

We can see that there are `nan` values in the `Bare_Nuclei` column.

In [17]:
# check how many nan values in `Bare_Nuclei` column

# 코드 8

16

We can see that there are 16 `nan` values in the dataset. I will impute missing values after dividing the dataset into training and test set.

### check frequency distribution of target variable `Class`

In [18]:
# view frequency distribution of values in `Class` variable

df['Class'].value_counts()

2    453
4    237
Name: Class, dtype: int64

### check  percentage of frequency distribution of `Class`

In [19]:
# view percentage of frequency distribution of values in `Class` variable

df['Class'].value_counts()/np.float(len(df))

2    0.656522
4    0.343478
Name: Class, dtype: float64

We can see that the `Class` variable contains 2 class labels - `2` and `4`. `2` stands for benign and `4` stands for malignant cancer.

### Outliers in numerical variables

In [20]:
# view summary statistics in numerical variables

print(round(df.describe(),2))

       Clump_thickness  Uniform_Cell_Size  Uniform_Cell_Shape  \
count           690.00             690.00              690.00   
mean              4.43               3.12                3.19   
std               2.83               3.03                2.96   
min               1.00               1.00                1.00   
25%               2.00               1.00                1.00   
50%               4.00               1.00                1.00   
75%               6.00               5.00                5.00   
max              10.00              10.00               10.00   

       Marginal_Adhesion  Single_Epithelial_Cell_Size  Bare_Nuclei  \
count             690.00                       690.00       674.00   
mean                2.81                         3.22         3.56   
std                 2.87                         2.22         3.66   
min                 1.00                         1.00         1.00   
25%                 1.00                         2.00         1.

kNN algorithm is robust to outliers.

# **3. Declare feature vector and target variable** <a class="anchor" id="9"></a>


In [21]:
X = df.drop(['Class'], axis=1)

y = df['Class']

# **4. Split data into separate training and test set** <a class="anchor" id="10"></a>



In [22]:
# split X and y into training and testing sets

from sklearn.model_selection import train_test_split

# 코드 9  (test_size = 0.2 사용하세요.)

In [23]:
# check the shape of X_train and X_test

X_train.shape, X_test.shape

((552, 9), (138, 9))

# **5. Feature Engineering** 


**Feature Engineering** is the process of transforming raw data into useful features that help us to understand our model better and increase its predictive power. 


In [24]:
# check data types in X_train

X_train.dtypes

Clump_thickness                  int64
Uniform_Cell_Size                int64
Uniform_Cell_Shape               int64
Marginal_Adhesion                int64
Single_Epithelial_Cell_Size      int64
Bare_Nuclei                    float64
Bland_Chromatin                  int64
Normal_Nucleo                    int64
Mitoses                          int64
dtype: object

### Engineering missing values in variables



In [25]:
# check missing values in numerical variables in X_train

X_train.isnull().sum()

Clump_thickness                 0
Uniform_Cell_Size               0
Uniform_Cell_Shape              0
Marginal_Adhesion               0
Single_Epithelial_Cell_Size     0
Bare_Nuclei                    12
Bland_Chromatin                 0
Normal_Nucleo                   0
Mitoses                         0
dtype: int64

In [26]:
# check missing values in numerical variables in X_test

X_test.isnull().sum()

Clump_thickness                0
Uniform_Cell_Size              0
Uniform_Cell_Shape             0
Marginal_Adhesion              0
Single_Epithelial_Cell_Size    0
Bare_Nuclei                    4
Bland_Chromatin                0
Normal_Nucleo                  0
Mitoses                        0
dtype: int64

In [27]:
# print percentage of missing values in the numerical variables in training set

for col in X_train.columns:
    if X_train[col].isnull().mean()>0:
        print(col, round(X_train[col].isnull().mean(),4))

Bare_Nuclei 0.0217


### Assumption


I assume that the data are missing completely at random (MCAR). There are two methods which can be used to impute missing values. One is mean or median imputation and other one is random sample imputation. When there are outliers in the dataset, we should use median imputation. So, I will use median imputation because median imputation is robust to outliers.


I will impute missing values with the appropriate statistical measures of the data, in this case median. Imputation should be done over the training set, and then propagated to the test set. It means that the statistical measures to be used to fill missing values both in train and test set, should be extracted from the train set only. This is to avoid overfitting.

In [28]:
# impute missing values in X_train and X_test with respective column median in X_train

for df1 in [X_train, X_test]:
    for col in X_train.columns:
        col_median=X_train[col].median()
        df1[col].fillna(col_median, inplace=True)           
      

In [29]:
# check again missing values in numerical variables in X_train

X_train.isnull().sum()

Clump_thickness                0
Uniform_Cell_Size              0
Uniform_Cell_Shape             0
Marginal_Adhesion              0
Single_Epithelial_Cell_Size    0
Bare_Nuclei                    0
Bland_Chromatin                0
Normal_Nucleo                  0
Mitoses                        0
dtype: int64

In [30]:
# check missing values in numerical variables in X_test

X_test.isnull().sum()

Clump_thickness                0
Uniform_Cell_Size              0
Uniform_Cell_Shape             0
Marginal_Adhesion              0
Single_Epithelial_Cell_Size    0
Bare_Nuclei                    0
Bland_Chromatin                0
Normal_Nucleo                  0
Mitoses                        0
dtype: int64

We can see that there are no missing values in X_train and X_test.

In [31]:
X_train.head()

,Clump_thickness,Uniform_Cell_Size,Uniform_Cell_Shape,Marginal_Adhesion,Single_Epithelial_Cell_Size,Bare_Nuclei,Bland_Chromatin,Normal_Nucleo,Mitoses
127,3,1,1,1,2,1.0,3,1,1
391,10,8,7,4,3,10.0,7,9,1
21,10,5,5,3,6,7.0,7,10,1
549,7,8,3,7,4,5.0,7,8,2
339,8,5,5,5,2,10.0,4,3,1


In [32]:
X_test.head()

,Clump_thickness,Uniform_Cell_Size,Uniform_Cell_Shape,Marginal_Adhesion,Single_Epithelial_Cell_Size,Bare_Nuclei,Bland_Chromatin,Normal_Nucleo,Mitoses
14,8,7,5,10,7,9.0,5,5,4
586,8,10,10,10,6,10.0,10,10,1
140,3,1,1,1,2,1.0,1,1,1
492,4,1,1,1,1,1.0,2,1,1
350,5,1,1,3,2,1.0,1,1,1


We now have training and testing set ready for model building. Before that, we should map all the feature variables onto the same scale. It is called `feature scaling`. I will do it as follows.

# **6. Feature Scaling** 

In [33]:
cols = X_train.columns

In [34]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train = scaler.fit_transform(X_train)

X_test = scaler.transform(X_test)


In [35]:
X_train = pd.DataFrame(X_train, columns=[cols])

In [36]:
X_test = pd.DataFrame(X_test, columns=[cols])

In [37]:
X_train.head()

,Clump_thickness,Uniform_Cell_Size,Uniform_Cell_Shape,Marginal_Adhesion,Single_Epithelial_Cell_Size,Bare_Nuclei,Bland_Chromatin,Normal_Nucleo,Mitoses
0,-0.477854,-0.690723,-0.731485,-0.626345,-0.543323,-0.671300,-0.156443,-0.608018,-0.330889
1,2.054968,1.638207,1.319894,0.423480,-0.092324,1.821240,1.520380,2.014059,-0.330889
2,2.054968,0.640094,0.636101,0.073538,1.260672,0.990393,1.520380,2.341818,-0.330889
3,0.969473,1.638207,-0.047692,1.473305,0.358675,0.436496,1.520380,1.686299,0.250801
4,1.331305,0.640094,0.636101,0.773422,-0.543323,1.821240,0.262763,0.047501,-0.330889


We now have `X_train` dataset ready to be fed into the Logistic Regression classifier. I will do it as follows.

# **7. Fit K Neighbours Classifier to the training set** 

In [38]:
# import KNeighbors ClaSSifier from sklearn
from sklearn.neighbors import KNeighborsClassifier


# instantiate the model
knn = KNeighborsClassifier(n_neighbors=3)


# fit the model to the training set
# 코드 10


KNeighborsClassifier(n_neighbors=3)

# **8. Predict test-set results** 

In [39]:
y_pred = knn.predict(X_test)

y_pred

array([4, 4, 2, 2, 2, 2, 2, 2, 4, 4, 4, 2, 2, 4, 4, 2, 4, 4, 2, 4, 2, 2,
       4, 2, 2, 2, 4, 2, 2, 2, 2, 2, 4, 4, 4, 4, 2, 2, 4, 2, 2, 4, 4, 2,
       4, 2, 4, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2, 4, 2, 4, 2, 2, 2, 2, 2, 4,
       4, 4, 2, 2, 2, 4, 4, 4, 4, 4, 2, 4, 2, 2, 2, 2, 4, 2, 2, 2, 2, 2,
       2, 4, 2, 2, 4, 2, 2, 4, 4, 4, 2, 2, 2, 2, 2, 2, 2, 4, 2, 2, 4, 4,
       4, 2, 4, 4, 2, 2, 2, 2, 4, 2, 2, 2, 2, 2, 4, 2, 4, 4, 4, 4, 2, 2,
       4, 2, 4, 2, 4, 2], dtype=int64)

### predict_proba method


**predict_proba** method gives the probabilities for the target variable(2 and 4) in this case, in array form.

`2 is for probability of benign cancer` and `4 is for probability of malignant cancer.`

In [40]:
# probability of getting output as 2 - benign cancer

knn.predict_proba(X_test)[:,0]

array([0.        , 0.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 0.33333333, 0.        ,
       0.        , 1.        , 1.        , 0.        , 0.        ,
       1.        , 0.        , 0.        , 1.        , 0.        ,
       1.        , 1.        , 0.        , 1.        , 1.        ,
       1.        , 0.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 0.        , 0.        , 0.        ,
       0.        , 1.        , 1.        , 0.        , 1.        ,
       1.        , 0.        , 0.        , 1.        , 0.33333333,
       1.        , 0.        , 1.        , 1.        , 1.        ,
       1.        , 1.        , 1.        , 1.        , 0.        ,
       1.        , 1.        , 0.        , 1.        , 0.        ,
       1.        , 1.        , 1.        , 1.        , 0.66666667,
       0.33333333, 0.        , 0.        , 1.        , 1.        ,
       1.        , 0.        , 0.        , 0.        , 0.     

In [41]:
# probability of getting output as 4 - malignant cancer

knn.predict_proba(X_test)[:,1]

array([1.        , 1.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.66666667, 1.        ,
       1.        , 0.        , 0.        , 1.        , 1.        ,
       0.        , 1.        , 1.        , 0.        , 1.        ,
       0.        , 0.        , 1.        , 0.        , 0.        ,
       0.        , 1.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 1.        , 1.        , 1.        ,
       1.        , 0.        , 0.        , 1.        , 0.        ,
       0.        , 1.        , 1.        , 0.        , 0.66666667,
       0.        , 1.        , 0.        , 0.        , 0.        ,
       0.        , 0.        , 0.        , 0.        , 1.        ,
       0.        , 0.        , 1.        , 0.        , 1.        ,
       0.        , 0.        , 0.        , 0.        , 0.33333333,
       0.66666667, 1.        , 1.        , 0.        , 0.        ,
       0.        , 1.        , 1.        , 1.        , 1.     

# **9. Check accuracy score** 

In [42]:
from sklearn.metrics import accuracy_score

print('Model accuracy score: {0:0.4f}'. format(accuracy_score(y_test, y_pred)))

Model accuracy score: 0.9565


Here, **y_test** are the true class labels and **y_pred** are the predicted class labels in the test-set.

### Compare the train-set and test-set accuracy


Now, I will compare the train-set and test-set accuracy to check for overfitting.

In [43]:
y_pred_train = knn.predict(X_train)

In [44]:
print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, y_pred_train)))

Training-set accuracy score: 0.9855


### Check for overfitting and underfitting

In [45]:
# print the scores on training and test set

print('Training set score: {:.4f}'.format(knn.score(X_train, y_train)))

print('Test set score: {:.4f}'.format(knn.score(X_test, y_test)))

Training set score: 0.9855
Test set score: 0.9565


The training-set accuracy score is 0.9821 while the test-set accuracy to be 0.9714. These two values are quite comparable. So, there is no question of overfitting. 


### Compare model accuracy with null accuracy


So, the model accuracy is 0.9714. But, we cannot say that our model is very good based on the above accuracy. We must compare it with the **null accuracy**. Null accuracy is the accuracy that could be achieved by always predicting the most frequent class.

So, we should first check the class distribution in the test set. 

In [46]:
# check class distribution in test set

y_test.value_counts()

2    88
4    50
Name: Class, dtype: int64

We can see that the occurences of most frequent class is 85. So, we can calculate null accuracy by dividing 85 by total number of occurences.

In [47]:
# check null accuracy score

null_accuracy = (85/(85+55))

print('Null accuracy score: {0:0.4f}'. format(null_accuracy))

Null accuracy score: 0.6071


We can see that our model accuracy score is 0.9714 but null accuracy score is 0.6071. So, we can conclude that our K Nearest Neighbors model is doing a very good job in predicting the class labels.